### PR TITLE
Finish adding types to `rosidl_parser`

### PR DESCRIPTION
**Pull request by InvincibleRMC**
Has to add lots of None checks inside the code. So if possible run the CI with downstream to make sure I didn't make a mistake anywhere.

### CODE REVIEW

**Review by sloretz — APPROVED** [review]

**Comment by sloretz** [comment]
Pulls: https://github.com/ros2/rosidl/pull/832
Gist: https://gist.githubusercontent.com/sloretz/d7cadee3ed1582742b70a09693d856be/raw/3a359eb7e3d161765b7fa09b02928204a941bb9e/ros2.repos
BUILD args:  --packages-above-and-dependencies rosidl_parser
TEST args:  --packages-above rosidl_parser
ROS Distro: rolling
Job: ci_launcher
ci_launcher ran: https://ci.ros2.org/job/ci_launcher/14758
* Linux [Build Status](http://ci.ros2.org/job/ci_linux/22061/)
* Linux-aarch64 [Build Status](http://ci.ros2.org/job/ci_linux-aarch64/16317/)
* Linux-rhel [Build Status](http://ci.ros2.org/job/ci_linux-rhel/1579/)
* Windows [Build Status](http://ci.ros2.org/job/ci_windows/22794/)

File: rosidl_parser/rosidl_parser/parser.py [hunk 1]
```diff
@@ -16,12 +16,23 @@
...
+from typing import Dict
+from typing import List
+from typing import Literal
+from typing import Optional
+from typing import TYPE_CHECKING
+from typing import Union
 
 from lark import Lark
 from lark.lexer import Token
+from lark.tree import Branch
```
[the remark this answers is not in the document]
Comment (InvincibleRMC) [reply rosidl_parser/rosidl_parser/parser.py]: Ok I added the `try/except`.

**Comment by sloretz** [comment]
RHEL rebuild: [Build Status](https://ci.ros2.org/job/ci_linux-rhel/1580/)

**Comment by InvincibleRMC** [comment]
Moved `ParseTree` into the try except as well.

**Comment by sloretz** [comment]
RHEL rebuild: [Build Status](https://ci.ros2.org/job/ci_linux-rhel/1583/)

**Review by fujitatomoya — APPROVED** [review]
lgtm with green CI. (including a couple of bug fixes!)

**Comment by InvincibleRMC** [comment]
I removed the try/except and just left the copied the definitions since the Mypy running on `Windows` and `rhel` CI does not support try/except yet. Support was added for this in 0.990 https://github.com/python/mypy/pull/13969.

**Review by sloretz — APPROVED** [review]

**Comment by sloretz** [comment]
Pulls: https://github.com/ros2/rosidl/pull/832
Gist: https://gist.githubusercontent.com/sloretz/ea75fed2da465bcae101db4a71a7237e/raw/3a359eb7e3d161765b7fa09b02928204a941bb9e/ros2.repos
BUILD args:  --packages-above-and-dependencies rosidl_parser
TEST args:  --packages-above rosidl_parser
ROS Distro: rolling
Job: ci_launcher
ci_launcher ran: https://ci.ros2.org/job/ci_launcher/14767
* Linux [Build Status](http://ci.ros2.org/job/ci_linux/22071/)
* Linux-aarch64 [Build Status](http://ci.ros2.org/job/ci_linux-aarch64/16328/)
* Linux-rhel [Build Status](http://ci.ros2.org/job/ci_linux-rhel/1590/)
* Windows [Build Status](http://ci.ros2.org/job/ci_windows/22805/)

**Comment by sloretz** [comment]
Checking latest commit on Windows: [Build Status](https://ci.ros2.org/job/ci_windows/22808/)

**Comment by sloretz** [comment]
@mergifyio update

**Comment by sloretz** [comment]
Pulls: https://github.com/ros2/rosidl/pull/832
Gist: https://gist.githubusercontent.com/sloretz/2c5962b66dd470fe6926c1dbc0f2d439/raw/3a359eb7e3d161765b7fa09b02928204a941bb9e/ros2.repos
BUILD args:  --packages-above-and-dependencies rosidl_parser
TEST args:  --packages-above rosidl_parser
ROS Distro: rolling
Job: ci_launcher
ci_launcher ran: https://ci.ros2.org/job/ci_launcher/14781
* Linux [Build Status](http://ci.ros2.org/job/ci_linux/22085/)
* Linux-aarch64 [Build Status](http://ci.ros2.org/job/ci_linux-aarch64/16342/)
* Linux-rhel [Build Status](http://ci.ros2.org/job/ci_linux-rhel/1604/)
* Windows [Build Status](http://ci.ros2.org/job/ci_windows/22824/)

**Comment by InvincibleRMC** [comment]
@sloretz  Do you need anything else from me?

**Comment by sloretz** [comment]
> @sloretz Do you need anything else from me?

Ooops, forgot about this one. Thank you for the PR!

**Pull request merged**